### PR TITLE
Flat field and pedestal modules

### PR DESCRIPTION
**Pull request by FrancaCassol**
This PR contain a first version of the flat field and pedestal modules.
It is compliant with the present master of ctapipe
It removes the LST event reader from cta-lstchain and uses the reader of ctapipe_io_lst 
It creates a calib/camera directory where all modules related to camera calibration are put
It creates a lstchain/tools directory with two examples of tools related to the writing of flat-field and pedestal data (these tools are meant as example, they are very preliminaries! )
It adds a notebook with some example on how the flat-field module and the flat-field tool works

In general, this code corresponds to "work in progress", so it will be subject to fast changes and developments, this PR is meant to start  the code sharing and to establish a common base for parallel and complementary code developments

### CODE REVIEW

**Comment by FrancaCassol** [comment]
I forgot to add that for the moment we don't have a test data file with flat-field and pedestal events, as soon it will be available, I will add it to cta-lstchain-extra and use it for the tests and examples

**Comment by rlopezcoto** [comment]
Thanks @FrancaCassol, a huge PR indeed, we'll have a look as soon as possible.

**Comment by FrancaCassol** [comment]
:-) 
Remember that the code is in fast evolution, so do not look too much at the details.
Also, concerting the data model, it is still in a preliminary version, I will add soon some Fields to make it more complete

**Comment by vuillaut** [comment]
Hi @FrancaCassol 
I see there are conflicting files now.
Also, Travis is up and running. Yeah!
So if you can add another commit to this PR (solving the conflicts), it should trigger a build (that won't pass because now we are setup with the ctapipe v0.6.1 but I'll change that quickly).

**Comment by rlopezcoto** [comment]
Some test modules are still not passing travis builds:
>  ImportError: cannot import name 'ChargeExtractor' from 'ctapipe.image'
Is ChargeExtractor already accepted in ctapipe?

**Comment by FrancaCassol** [comment]
Yes, it should be, anyway for me it works...
I don't understand where is the problem. If you clone the PR is it working for you?

**Comment by vuillaut** [comment]
Hi.
We are still not cloning ctapipe master for the travis build.
Sorry for that, I haven't found the time to do it since last Friday.
I'll look into it right now.

File: lstchain/calib/camera/tests/test_pedestals.py [hunk 1]
```diff
@@ -0,0 +1,46 @@
+from ctapipe.utils import get_dataset_path
+
+import numpy as np
+from ctapipe.io import event_source
+from .. import PedestalIntegrator
+from .. import pedestals
+
+input_reader = event_source("/ctadata/franca/LST/LST-1.4.Run00167.0001.fits.fz", max_events=10)
```
Comment (vuillaut) [inline lstchain/calib/camera/tests/test_pedestals.py]: This needs to be changed to use an accessible file (prob. in `cta-lstchain-extra`)

Comment (FrancaCassol) [reply lstchain/calib/camera/tests/test_pedestals.py]: Yes, the problem is that I miss a data file small enough to be put into the extra directory, I asked already for it, let ask again.

Comment (vuillaut) [reply lstchain/calib/camera/tests/test_pedestals.py]: Ok, thank you.
In this state, the tests would never pass.

Comment (FrancaCassol) [reply lstchain/calib/camera/tests/test_pedestals.py]: Should we suspend it for the moment?

Comment (vuillaut) [reply lstchain/calib/camera/tests/test_pedestals.py]: If the tests are passing on your computer and this module is important and we don't want anymore, maybe we can comment the tests. In this case open an issue to reactivate tests later and solve it asap.

**Comment by vuillaut** [comment]
Note that I haven't tested the rest of the code.
I don't know if @rlopezcoto had time to give it a full review?

**Comment by FrancaCassol** [comment]
Yes, don't forget that the code is in development. I am working on it.

**Comment by rlopezcoto** [comment]
@vuillaut: not yet, I was waiting for it to solve incompatibilities and pass Travis tests to start a deep review.
@FrancaCassol: the only thing we need for the code to pass is that one file? (I guess it contains pedestal events,right?) It may be too heavy for cta-lstchain-extra so we may need to find another solution. I also suggest to suspend the test for the moment

**Comment by FrancaCassol** [comment]
As said I asked the camera people to run a short run in order to get such a file. But for the moment will suspend this test. 
In order to get this PR accepted we need also the new version of ctapipe_io_lst.

**Comment by rlopezcoto** [comment]
Hi @FrancaCassol, whenever https://github.com/cta-observatory/ctapipe_io_lst/pull/3
is merged and #103 is accepted (that will take the master version of ctapipe_io_lst), there should not be any more problems with Travis checks, right?

**Comment by FrancaCassol** [comment]
Yes, in principle, all should go fine. But I must still do two PRs

**Comment by FrancaCassol** [comment]
Hi, is it normal that this problem is still there?

**Comment by rlopezcoto** [comment]
which problem @FrancaCassol? I now see a conflict with `lstchain/io/tests/test_lsteventsource.py`

**Comment by FrancaCassol** [comment]
I don't see the conflict (where can you see it?). But Let me verify, I am indeed changing the code and that could have happened

**Comment by rlopezcoto** [comment]
Screen Shot 2019-03-14 at 14 24 22

**Comment by FrancaCassol** [comment]
My question was how to see from this page which are the conflict, but, now I understand, the point is that we should first accept the ctapipe_io_lst PR.
But, I propose to do it as soon as I finish to integrate the outliers in the data model (I am almost finished), is it ok? The this I will update also this PR and all should go fine

**Comment by rlopezcoto** [comment]
Hi Franca, sorry, I thought you could not see them. 

> I propose to do it as soon as I finish to integrate the outliers in the data model 

Do you mean into `ctapipe_io_lst`? It should be fine, please go ahead

**Comment by FrancaCassol** [comment]
Hi, it seems that finally this PR do not have conflicts. Do we want to proceed for a merge?

**Comment by rlopezcoto** [comment]
Hi @FrancaCassol, thanks for solving all the conflicts and leave the PR ready to be merged. I would like now to have a carful look to the code. This is an important and long PR, so I guess somebody else (@vuillaut?) will also have a look and send suggestions, would that be ok with you?

**Comment by FrancaCassol** [comment]
Surely, I thought you did already ;-) 
please suggest any improvement you think. 
Since a lot of things are still missing in the meanwhile I will continue to develop

**Comment by FrancaCassol** [comment]
@rlopezcoto it seems ok now . Please, remember that this is a code under development, it must/can not be perfect.

**Comment by FrancaCassol** [comment]
Hi @rlopezcoto 

> * It would be nice to get a small pedestal file in cta-lstchain-extra for testing

I asked already several times to the DAQ people to get a small file with flatfield and pedestal events, unfortunately it seems not so easy to produce. The present files are all too big to be added to the extra directory.

> * I understand that the code for broken pixels and other fine-tuning corrections is going to be added afterwards, right? We can leave without that for the moment.

Actually I am almost ready also with that, this will be my next PR as soon this one is accepted and @Pawel do a PR with the last R0.py version (see [https://github.com/cta-observatory/cta-lstchain/issues/106](https://github.com/cta-observatory/cta-lstchain/issues/106) )

> * Eliminating lsteventsource.py is likely going to break some code, have you considered #99 in the new definition on of LSTEventSource? (also, sorry that I did not follow in deep the discussions but what was the reason to change it to LSTEventContainer)? I am a bit worried about backwards compatibility of this whole calls.

This move is asked by the refactoring of the eventsource module in ctapipe. Now everything is in ctapipe_io_lst. It is in general a good change. 

> * ` notebooks/Read_and_plot_LST_Data.ipynb` will not work after eliminating the definition of LSTEventSource, could you please modify this and any other call to this class?

This notebook can not work already, with the present ctapipe, I will change it

> * You are only introducing one method to calculate the pedestal averaging over the full window, right? (I may be missing something within the code, sorry)

No, the pedestal can be calculated with the charge integrator that you like (it is an argument of the module). In my next PR I will provide a notebook for that.

It would be nice to have all this and the next PR accepted for the ASWG meeting (this will probably not be the best code but at least  a starting point)

**Comment by rlopezcoto** [comment]
Thanks Franca, I also believe that this new arrangement is good, but we need to adapt the code to it. I made a PR in your fork solving other LSTEventSource calls:
https://github.com/FrancaCassol/cta-lstchain/pull/2

Also, I found some `LSTEventSource` printing in `calculate_flatfield.ipynb`, I hope that is not a problem (could you please check very quickly?). After you accept and merge the previous PR with you branch, if nobody opposes, I'll merge it with the master.

**Comment by FrancaCassol** [comment]
Hi @rlopezcoto,

I think you can accept this PR now. Then, there will be soon an other PR with an updated version of the code that provides also the outliers from flatfield and pedestal events

**Comment by rlopezcoto** [comment]
Thanks @FrancaCassol, I'll merge.

**Pull request merged**